<p style="text-align:center">
    <a href="https://skills.network/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkPY0220ENSkillsNetwork900-2022-01-01" target="_blank">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="200" alt="Skills Network Logo">
    </a>
</p>


<h1>Extracting and Visualizing Stock Data</h1>
<h2>Description</h2>


Extracting essential data from a dataset and displaying it is a necessary part of data science; therefore individuals can make correct decisions based on the data. In this assignment, you will extract some stock data, you will then display this data in a graph.


<h2>Table of Contents</h2>
<div class="alert alert-block alert-info" style="margin-top: 20px">
    <ul>
        <li>Define a Function that Makes a Graph</li>
        <li>Question 1: Use yfinance to Extract Stock Data</li>
        <li>Question 2: Use Webscraping to Extract Tesla Revenue Data</li>
        <li>Question 3: Use yfinance to Extract Stock Data</li>
        <li>Question 4: Use Webscraping to Extract GME Revenue Data</li>
        <li>Question 5: Plot Tesla Stock Graph</li>
        <li>Question 6: Plot GameStop Stock Graph</li>
    </ul>
<p>
    Estimated Time Needed: <strong>30 min</strong></p>
</div>

<hr>


***Note***:- If you are working in IBM Cloud Watson Studio, please replace the command for installing nbformat from `!pip install nbformat==4.2.0` to simply `!pip install nbformat`


In [39]:
!pip install yfinance==0.1.67
!mamba install bs4==4.10.0 -y
!pip install nbformat==5.2.0


                  __    __    __    __
                 /  \  /  \  /  \  /  \
                /    \/    \/    \/    \
███████████████/  /██/  /██/  /██/  /████████████████████████
              /  / \   / \   / \   / \  \____
             /  /   \_/   \_/   \_/   \    o \__,
            / _/                       \_____/  `
            |/
        ███╗   ███╗ █████╗ ███╗   ███╗██████╗  █████╗
        ████╗ ████║██╔══██╗████╗ ████║██╔══██╗██╔══██╗
        ██╔████╔██║███████║██╔████╔██║██████╔╝███████║
        ██║╚██╔╝██║██╔══██║██║╚██╔╝██║██╔══██╗██╔══██║
        ██║ ╚═╝ ██║██║  ██║██║ ╚═╝ ██║██████╔╝██║  ██║
        ╚═╝     ╚═╝╚═╝  ╚═╝╚═╝     ╚═╝╚═════╝ ╚═╝  ╚═╝

        mamba (1.4.2) supported by @QuantStack

        GitHub:  https://github.com/mamba-org/mamba
        Twitter: https://twitter.com/QuantStack

█████████████████████████████████████████████████████████████


Looking for: ['bs4==4.10.0']

[+] 0.0s
pkgs/main/linux-64 ━━━━╸━━━━━━━━━━━━━━━╸━━━━   0.0 B /  ??.?MB @  ??.?MB/s

In [40]:
import yfinance as yf
import pandas as pd
import requests
from bs4 import BeautifulSoup
!pip install html5lib
import plotly.graph_objects as go
from plotly.subplots import make_subplots
!pip install --upgrade beautifulsoup4

In Python, you can ignore warnings using the warnings module. You can use the filterwarnings function to filter or ignore specific warning messages or categories.


In [41]:
import warnings
# Ignore all warnings
warnings.filterwarnings("ignore", category=FutureWarning)

## Define Graphing Function


In this section, we define the function `make_graph`. You don't have to know how the function works, you should only care about the inputs. It takes a dataframe with stock data (dataframe must contain Date and Close columns), a dataframe with revenue data (dataframe must contain Date and Revenue columns), and the name of the stock.


In [42]:
def make_graph(stock_data, revenue_data, stock):
    fig = make_subplots(rows=2, cols=1, shared_xaxes=True, subplot_titles=("Historical Share Price", "Historical Revenue"), vertical_spacing = .3)
    stock_data_specific = stock_data[stock_data.Date <= '2021--06-14']
    revenue_data_specific = revenue_data[revenue_data.Date <= '2021-04-30']
    fig.add_trace(go.Scatter(x=pd.to_datetime(stock_data_specific.Date, infer_datetime_format=True), y=stock_data_specific.Close.astype("float"), name="Share Price"), row=1, col=1)
    fig.add_trace(go.Scatter(x=pd.to_datetime(revenue_data_specific.Date, infer_datetime_format=True), y=revenue_data_specific.Revenue.astype("float"), name="Revenue"), row=2, col=1)
    fig.update_xaxes(title_text="Date", row=1, col=1)
    fig.update_xaxes(title_text="Date", row=2, col=1)
    fig.update_yaxes(title_text="Price ($US)", row=1, col=1)
    fig.update_yaxes(title_text="Revenue ($US Millions)", row=2, col=1)
    fig.update_layout(showlegend=False,
    height=900,
    title=stock,
    xaxis_rangeslider_visible=True)
    fig.show()

## Question 1: Use yfinance to Extract Stock Data


Using the `Ticker` function enter the ticker symbol of the stock we want to extract data on to create a ticker object. The stock is Tesla and its ticker symbol is `TSLA`.


In [43]:
tesla=yf.Ticker("TSLA")

Using the ticker object and the function `history` extract stock information and save it in a dataframe named `tesla_data`. Set the `period` parameter to `max` so we get information for the maximum amount of time.


In [44]:
tesla_data=tesla.history(period="max")


**Reset the index** using the `reset_index(inplace=True)` function on the tesla_data DataFrame and display the first five rows of the `tesla_data` dataframe using the `head` function. Take a screenshot of the results and code from the beginning of Question 1 to the results below.


In [45]:
tesla_data.reset_index(inplace=True)
tesla_data.head()

Date      Open      High       Low     Close     Volume  Dividends  \
0 2010-06-29  1.266667  1.666667  1.169333  1.592667  281494500          0   
1 2010-06-30  1.719333  2.028000  1.553333  1.588667  257806500          0   
2 2010-07-01  1.666667  1.728000  1.351333  1.464000  123282000          0   
3 2010-07-02  1.533333  1.540000  1.247333  1.280000   77097000          0   
4 2010-07-06  1.333333  1.333333  1.055333  1.074000  103003500          0   

   Stock Splits  
0           0.0  
1           0.0  
2           0.0  
3           0.0  
4           0.0

## Question 2: Use Webscraping to Extract Tesla Revenue Data


Use the `requests` library to download the webpage https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/revenue.htm Save the text of the response as a variable named `html_data`.


In [61]:
import requests

# URL of the webpage to download
url = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/revenue.htm"

# Send a GET request to the URL
response = requests.get(url)

# Check if the request was successful (status code 200)
if response.status_code == 200:
    # Save the text of the response as html_data
    html_data = response.text
    print("Webpage downloaded successfully.")
else:
    print(f"Failed to download webpage. Status code: {response.status_code}")



Webpage downloaded successfully.


Parse the html data using `beautiful_soup`.


In [62]:
from bs4 import BeautifulSoup

# Assuming html_data contains the HTML content obtained from the webpage
# Parse the HTML using BeautifulSoup
soup = BeautifulSoup(html_data, 'html.parser')

# Now you can work with the parsed HTML content using BeautifulSoup
# For example, you can find elements by tag name, class name, etc.

# Example: Find all <a> tags in the parsed HTML
all_links = soup.find_all('a')

# Example: Find the title of the webpage
title = soup.title

# Example: Extract text from a specific element
paragraph_text = soup.find('p').text

# You can perform various operations on the parsed HTML using BeautifulSoup.


Using `BeautifulSoup` or the `read_html` function extract the table with `Tesla Revenue` and store it into a dataframe named `tesla_revenue`. The dataframe should have columns `Date` and `Revenue`.


<details><summary>Click here if you need help locating the table</summary>

```
    
Below is the code to isolate the table, you will now need to loop through the rows and columns like in the previous lab
    
soup.find_all("tbody")[1]
    
If you want to use the read_html function the table is located at index 1

We are focusing on quarterly revenue in the lab.
```

</details>


In [63]:
from bs4 import BeautifulSoup
import pandas as pd

# Assuming html_data contains the HTML content obtained from the webpage
soup = BeautifulSoup(html_data, 'html.parser')

# Find all tables in the HTML
tables = soup.find_all('table')

# Look through each table to find the one containing Tesla Revenue
tesla_revenue_table = None
for table in tables:
    if 'Tesla Revenue' in table.text:
        tesla_revenue_table = table
        break

if tesla_revenue_table:
    # Extract table headers
    headers = tesla_revenue_table.find_all('th')
    columns = [header.text.strip() for header in headers]

    # Extract table rows
    rows = tesla_revenue_table.find_all('tr')[1:]  # Exclude header row

    # Initialize lists to store data
    dates = []
    revenues = []

    # Extract data from rows
    for row in rows:
        data = row.find_all('td')
        date = data[0].text.strip()
        revenue = data[1].text.strip()
        dates.append(date)
        revenues.append(revenue)

    # Create DataFrame
    tesla_revenue = pd.DataFrame({
        'Date': dates,
        'Revenue': revenues
    })

    # Display the DataFrame
    print(tesla_revenue)








                             Date Revenue
0  Tesla Revenue 2010-2022 | TSLA        
1                     Macrotrends        
2                          Source        


Execute the following line to remove the comma and dollar sign from the `Revenue` column. 


In [64]:
# Assuming 'tesla_revenue' DataFrame has been created
tesla_revenue['Revenue'] = tesla_revenue['Revenue'].str.replace(',', '').str.replace('$', '')

# Convert the 'Revenue' column to numeric type
tesla_revenue['Revenue'] = pd.to_numeric(tesla_revenue['Revenue'])

# Display the DataFrame
print(tesla_revenue)


                             Date  Revenue
0  Tesla Revenue 2010-2022 | TSLA      NaN
1                     Macrotrends      NaN
2                          Source      NaN


Execute the following lines to remove an null or empty strings in the Revenue column.


In [57]:
#tesla_revenue.dropna(inplace=True)

#tesla_revenue = tesla_revenue[tesla_revenue['Revenue'] != ""]

Display the last 5 row of the `tesla_revenue` dataframe using the `tail` function. Take a screenshot of the results.


In [66]:
print(tesla_revenue.tail())

                             Date  Revenue
0  Tesla Revenue 2010-2022 | TSLA      NaN
1                     Macrotrends      NaN
2                          Source      NaN


## Question 3: Use yfinance to Extract Stock Data


Using the `Ticker` function enter the ticker symbol of the stock we want to extract data on to create a ticker object. The stock is GameStop and its ticker symbol is `GME`.


In [26]:
gamestop = yf.Ticker("GME")


Using the ticker object and the function `history` extract stock information and save it in a dataframe named `gme_data`. Set the `period` parameter to `max` so we get information for the maximum amount of time.


In [27]:
gme_data=tesla.history(period="max")


**Reset the index** using the `reset_index(inplace=True)` function on the gme_data DataFrame and display the first five rows of the `gme_data` dataframe using the `head` function. Take a screenshot of the results and code from the beginning of Question 3 to the results below.


In [28]:
gme_data.reset_index(inplace=True)
gme_data.head()

Date      Open      High       Low     Close     Volume  Dividends  \
0 2010-06-29  1.266667  1.666667  1.169333  1.592667  281494500          0   
1 2010-06-30  1.719333  2.028000  1.553333  1.588667  257806500          0   
2 2010-07-01  1.666667  1.728000  1.351333  1.464000  123282000          0   
3 2010-07-02  1.533333  1.540000  1.247333  1.280000   77097000          0   
4 2010-07-06  1.333333  1.333333  1.055333  1.074000  103003500          0   

   Stock Splits  
0           0.0  
1           0.0  
2           0.0  
3           0.0  
4           0.0

## Question 4: Use Webscraping to Extract GME Revenue Data


Use the `requests` library to download the webpage https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/stock.html. Save the text of the response as a variable named `html_data`.


In [67]:
import requests

# URL of the webpage to download
url = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/stock.html"

# Send a GET request to the URL
response = requests.get(url)

# Check if the request was successful (status code 200)
if response.status_code == 200:
    # Save the text of the response as html_data
    html_data = response.text
    print("Webpage downloaded successfully.")
else:
    print(f"Failed to download webpage. Status code: {response.status_code}")


Webpage downloaded successfully.


Parse the html data using `beautiful_soup`.


In [68]:
from bs4 import BeautifulSoup

# Assuming html_data contains the HTML content obtained from the webpage
soup = BeautifulSoup(html_data, 'html.parser')

# Now you can work with the parsed HTML content using BeautifulSoup
# For example, you can find elements by tag name, class name, etc.

# Example: Find all <a> tags in the parsed HTML
all_links = soup.find_all('a')

# Example: Find the title of the webpage
title = soup.title

# Example: Extract text from a specific element
paragraph_text = soup.find('p').text

# You can perform various operations on the parsed HTML using BeautifulSoup.


Using `BeautifulSoup` or the `read_html` function extract the table with `GameStop Revenue` and store it into a dataframe named `gme_revenue`. The dataframe should have columns `Date` and `Revenue`. Make sure the comma and dollar sign is removed from the `Revenue` column using a method similar to what you did in Question 2.


<details><summary>Click here if you need help locating the table</summary>

```
    
Below is the code to isolate the table, you will now need to loop through the rows and columns like in the previous lab
    
soup.find_all("tbody")[1]
    
If you want to use the read_html function the table is located at index 1


```

</details>


In [69]:
from bs4 import BeautifulSoup
import pandas as pd

# Assuming html_data contains the HTML content obtained from the webpage
soup = BeautifulSoup(html_data, 'html.parser')

# Find all tables in the HTML
tables = soup.find_all('table')

# Initialize variables to store the table containing GameStop Revenue data
gme_revenue_table = None

# Look through each table to find the one containing GameStop Revenue
for table in tables:
    if 'GameStop Revenue' in table.text:
        gme_revenue_table = table
        break

# If GameStop Revenue table is found
if gme_revenue_table:
    # Extract table headers
    headers = gme_revenue_table.find_all('th')
    columns = [header.text.strip() for header in headers]

    # Extract table rows
    rows = gme_revenue_table.find_all('tr')[1:]  # Exclude header row

    # Initialize lists to store data
    dates = []
    revenues = []

    # Extract data from rows
    for row in rows:
        data = row.find_all('td')
        date = data[0].text.strip()
        revenue = data[1].text.strip().replace(',', '').replace('$', '')  # Remove comma and dollar sign
        dates.append(date)
        revenues.append(revenue)

    # Create DataFrame
    gme_revenue = pd.DataFrame({
        'Date': dates,
        'Revenue': revenues
    })

    # Display the DataFrame
    print(gme_revenue)
else:
    print("GameStop Revenue table not found in the HTML.")


                               Date Revenue
0  GameStop Revenue 2006-2020 | GME        
1                       Macrotrends        
2                            Source        


Display the last five rows of the `gme_revenue` dataframe using the `tail` function. Take a screenshot of the results.


In [70]:
gme_revenue.tail()

Date Revenue
0  GameStop Revenue 2006-2020 | GME        
1                       Macrotrends        
2                            Source

## Question 5: Plot Tesla Stock Graph


Use the `make_graph` function to graph the Tesla Stock Data, also provide a title for the graph. The structure to call the `make_graph` function is `make_graph(tesla_data, tesla_revenue, 'Tesla')`. Note the graph will only show data upto June 2021.


In [74]:
import matplotlib.pyplot as plt
import pandas as pd
def make_graph(stock_data, revenue_data, title):
   
    if not isinstance(stock_data, pd.DataFrame):
        stock_data = pd.DataFrame(stock_data, columns=['Date', 'Close'])
        stock_data['Date'] = pd.to_datetime(stock_data['Date'])
    if not isinstance(revenue_data, pd.DataFrame):
        revenue_data = pd.DataFrame(revenue_data, columns=['Date', 'Revenue'])
        revenue_data['Date'] = pd.to_datetime(revenue_data['Date'])
    plt.figure(figsize=(10, 6))
    plt.plot(stock_data['Date'], stock_data['Close'], label='Stock Data')
    plt.bar(revenue_data['Date'], revenue_data['Revenue'], label='Revenue Data', color='orange')
    plt.title(f'Stock Data and Revenue for {title}')
    plt.xlabel('Date')
    plt.ylabel('Value')
    plt.xlim(stock_data['Date'].min(), pd.to_datetime('2021-06-30'))
    plt.xticks(rotation=45)
    plt.legend()

    plt.tight_layout()
    plt.show()
make_graph(tesla_data, tesla_revenue, 'Tesla')



ValueError: Shape of passed values is (1, 1), indices imply (1, 2)

## Question 6: Plot GameStop Stock Graph


Use the `make_graph` function to graph the GameStop Stock Data, also provide a title for the graph. The structure to call the `make_graph` function is `make_graph(gme_data, gme_revenue, 'GameStop')`. Note the graph will only show data upto June 2021.


In [75]:
import matplotlib.pyplot as plt

def make_graph(stock_data, revenue_data, company_name):
    # Extracting dates and stock prices
    dates = [data[0] for data in stock_data]
    prices = [data[1] for data in stock_data]

    revenue_dates = [data[0] for data in revenue_data]
    revenues = [data[1] for data in revenue_data]
    plt.figure(figsize=(10, 6))
    plt.plot(dates, prices, label='Stock Price', color='blue')
    plt.plot(revenue_dates, revenues, label='Revenue', color='green')
    plt.title(f'{company_name} Stock Data and Revenue')
    plt.xlabel('Date')
    plt.ylabel('Price/Revenue')
    plt.legend()
    plt.grid(True)
    plt.show()
gme_data = [...]  # Your GameStop stock data up to June 2021
gme_revenue = [...]  # Your GameStop revenue data up to June 2021
make_graph(gme_data, gme_revenue, 'GameStop')


TypeError: 'ellipsis' object is not subscriptable

<h2>About the Authors:</h2> 

<a href="https://www.linkedin.com/in/joseph-s-50398b136/">Joseph Santarcangelo</a> has a PhD in Electrical Engineering, his research focused on using machine learning, signal processing, and computer vision to determine how videos impact human cognition. Joseph has been working for IBM since he completed his PhD.

Azim Hirjani


## Change Log

| Date (YYYY-MM-DD) | Version | Changed By    | Change Description        |
| ----------------- | ------- | ------------- | ------------------------- |
| 2022-02-28        | 1.2     | Lakshmi Holla | Changed the URL of GameStop |
| 2020-11-10        | 1.1     | Malika Singla | Deleted the Optional part |
| 2020-08-27        | 1.0     | Malika Singla | Added lab to GitLab       |

<hr>

## <h3 align="center"> © IBM Corporation 2020. All rights reserved. <h3/>

<p>
